File: code/classes/downcasting.java
// Downcasting in Java
class Animal {
    public void makeSound() {
        System.out.println("Some sound");
    }
}

class Dog extends Animal {
    public void makeSound() {
        System.out.println("Woof");
    }
    
    public void fetch() {
        System.out.println("Fetching...");
    }
}

public class Main {
    public static void main(String[] args) {
        Animal animal = new Dog();
        
        // Downcasting
        if (animal instanceof Dog) {
            Dog dog = (Dog) animal;
            dog.makeSound();
            dog.fetch();
        }
    }
}

File: code/classes/downcasting.ts
// Downcasting in TypeScript
class Animal {
    makeSound(): void {
        console.log("Some sound");
    }
}

class Dog extends Animal {
    makeSound(): void {
        console.log("Woof");
    }
    
    fetch(): void {
        console.log("Fetching...");
    }
}

const animal: Animal = new Dog();

// Downcasting with type guard
if (animal instanceof Dog) {
    animal.makeSound();
    animal.fetch();
}

// Or using type assertion (with caution)
const dog = animal as Dog;
dog.makeSound();
dog.fetch();